### ISSUE TITLE
Only one of several TIdIPMCastClients is able to set to Active

### ISSUE THREAD

**Issue by Blackbird700** [IndySockets/Indy#419]
I was wondering why I'm not able to enable several of the TIdIPMCastClients?
As soon as I try to enable a second one it fails without an exception. 
(the active value is false even I just tried to set it to true)
If I disable the first one I can enable the second one and the second one does receive the multicast just fine.
Any hint would be appriciated. 
Thanks

**Comment by rlebeau** [comment]
Please provide an example. How are you configuring the clients? What does your code look like? Since `TIdIPMCastClient` uses listening sockets, you can't have multiple clients bound to the same local IP/Port at the same time. Is that what you are trying to do?

Also be aware that I just opened #422 related to this.

**Comment by Blackbird700** [comment]
Essentially the clients are the similar beside they supposed to listen to different multicast IP/Port.
Now, I did try to change the Bindings to, different IP and even different Port. (Two different NIC’s)
As you might can see in the attached test application.
But that doesn’t help. So guess I’m still something missing.
Thanks!

Daniel

**Comment by Blackbird700** [comment]
Interesting… just tried once again with different Bindings… This time it seems to work.
Seems like I need to dig more what I might did to wrong.

***@***.***


From: Daniel Ruoss
Sent: Sunday, July 24, 2022 1:32 PM
To: 'IndySockets/Indy' ***@***.***>; IndySockets/Indy ***@***.***>
Cc: Author ***@***.***>
Subject: RE: [IndySockets/Indy] Only one of several TIdIPMCastClients is able to set to Active (Issue #419)

Essentially the clients are the similar beside they supposed to listen to different multicast IP/Port.
Now, I did try to change the Bindings to, different IP and even different Port. (Two different NIC’s)
As you might can see in the attached test application.
But that doesn’t help. So guess I’m still something missing.
Thanks!

Daniel

**Comment by Blackbird700** [comment]
There is strange issue with setting the default port and then changing on the fly.
If I just set Client1.Port:=NewPort it doesn’t take it to the binding on first try only after “restart”.
So now, when I do this
Client1.Port:=NewPort
Client1.Bindings[0].Port:=NewPort
This seems to work more reliable.
At least with that so far it seems to be working as I needed it.
Thanks so much for your hints.
Daniel

From: Daniel Ruoss
Sent: Sunday, July 24, 2022 1:45 PM
To: 'IndySockets/Indy' ***@***.***>; 'IndySockets/Indy' ***@***.***>
Cc: 'Author' ***@***.***>
Subject: RE: [IndySockets/Indy] Only one of several TIdIPMCastClients is able to set to Active (Issue #419)

Interesting… just tried once again with different Bindings… This time it seems to work.
Seems like I need to dig more what I might did to wrong.

***@***.***


From: Daniel Ruoss
Sent: Sunday, July 24, 2022 1:32 PM
To: 'IndySockets/Indy' ***@***.******@***.***>>; IndySockets/Indy ***@***.******@***.***>>
Cc: Author ***@***.******@***.***>>
Subject: RE: [IndySockets/Indy] Only one of several TIdIPMCastClients is able to set to Active (Issue #419)

Essentially the clients are the similar beside they supposed to listen to different multicast IP/Port.
Now, I did try to change the Bindings to, different IP and even different Port. (Two different NIC’s)
As you might can see in the attached test application.
But that doesn’t help. So guess I’m still something missing.
Thanks!

Daniel

**Comment by rlebeau** [comment]
You can't join clients to multiple multicast groups using the same local listening IP/Port `Binding`.  A single `TIdPMCastClient` can join the same group on different local `Binding` IP/Ports, or multiple `TIdIPMCastClient`s can join different groups on different local `Binding` IP/Ports.

There is no `Port` property in `TIdIPMCastClient` (well, ok there is, but it is `protected`, and not actually used for anything in `TIdIPMCastClient`).  Changing the client's `DefaultPort` property is just a default (hence the name).  It is not intended to update any existing `Binding`, only provide a default whenever a new `Binding` is created.  So yes, if you want to update an existing `Binding`, you have to deactivate the client and update the `Binding` directly.  Other Indy components that have `Bindings` work this same way.

**Comment by Blackbird700** [comment]
I’m very well aware of that. (can’t join multiple groups using same local ip)
But the issue is this:
If you do this:
IdIPMCastClient.DefaultPort:=8001;
IdIPMCastClient.Active:=True;

Wait a while then

IdIPMCastClient.Active:=False;
IdIPMCastClient.DefaultPort:=8002;
IdIPMCastClient.Active:=True;

The binding does NOT change.
In other words, it’s still listing on 8001 instead 8002 that was my issue.
I have fixed it now by doing this:
IdIPMCastClient.DefaultPort:=8001;
IdIPMCastClient.Active:=True;
Wait a while
then
IdIPMCastClient.Active:=False;
IdIPMCastClient.DefaultPort:=8002;
if IdIPMCastClient.Bindings.Count>0 then
begin
      IdIPMCastClient.Bindings[0].Port:=IdIPMCastClient.DefaultPort;
end;
IdIPMCastClient.Active:=True;

NOW it works correct.

By the way the EXACT same issue/work around, I had to do IdHTTPServer as well.

Daniel

**Comment by rlebeau** [comment]
> the issue is this:
> If you do this:
> ...
> The binding does NOT change.
> In other words, it’s still listing on 8001 instead 8002 that was my issue.

It is NOT supposed to change.  You are just updating the *default* port, not the *active* port.  The existing `Bindings` entries are the *active* ports, and they can be reused as-is regardless of the new *default*.  If you want the new `DefaultPort` to take effect, you have to `Clear()` the `Bindings` first, eg:

```
IdIPMCastClient.Active := False;
IdIPMCastClient.Bindings.Clear; // <-- add this
IdIPMCastClient.DefaultPort := 8002;
IdIPMCastClient.Active := True;
```

Or, as you already discovered, you would have to alternatively update each `Binding.Port` directly without `Clear()`'ing the whole `Bindings` collection.

> I have fixed it now by doing this:
> ...

Yes, that will work (though, there could be more then 1 entry in the `Bindings`, so you should use a loop instead).

> By the way the EXACT same issue/work around, I had to do IdHTTPServer as well.

All of Indy's multi-binding listening components work this same way.  This is not a bug, it is part of their design.

**Comment by Blackbird700** [comment]
I just tried your suggestion with
IdIPMCastClient.Bindings.Clear
If I do that, I can’t connect anymore to the previous port NOR to the new port.
So that way certainly does not work
But thanks for the hint about I might should make a loop. Going to add that.
Daniel

**Comment by rlebeau** [comment]
> I just tried your suggestion with
> IdIPMCastClient.Bindings.Clear
> If I do that, I can’t connect anymore to the previous port NOR to the new port.
> So that way certainly does not work

It should be working. The `Active` setter should see the `Bindings` is empty and automatically add a new `Binding` object with its `Port` set to the current `DefaultPort`, and then open that `Binding` for listening.  Are you getting any errors raised from the `Active` setter in this situation?

**Comment by Blackbird700** [comment]
No, there is no exception and it even says enabled, when I check the Active value.
Daniel

**Comment by rlebeau** [comment]
On the first 1st listen, were you configuring any local `Binding.IP`?  On the 2nd listen, if you `Clear()` the `Bindings`, are you specifying a local `Binding.IP`?  If you just loop through the `Bindings` updating each `Binding.Port` directly, the previous `Binding.IP` is preserved.  Maybe that is a factor in your issue?
